dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +253,282 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
-                }
+            auto scopeExitToCleanRegistrations{ wil::scope_exit(
+            [&](){
+                    auto lock{ m_lock.lock_exclusive() };
 
-                winrt::hstring taskClsidStr{ winrt::to_hstring(taskClsid) };
-                winrt::hstring backgroundTaskFullName{ backgroundTaskName + taskClsidStr };
+                    m_comActivatorRegistration.reset();
 
-                auto tasks{ BackgroundTaskRegistration::AllTasks() };
-                bool isTaskRegistered{ std::any_of(std::begin(tasks), std::end(tasks),
-                    [&](auto&& task)
+                    // Clean the task registration only if it was created during this call
+                    if (m_pushTriggerRegistration)
                     {
-                        auto name{ task.Value().Name() };
-
-                        if (std::wstring_view(name).substr(0, backgroundTaskName.size()) != backgroundTaskName)
-                        {
-                            return false;
-                        }
+                        m_pushTriggerRegistration.Unregister(true);
+                        m_pushTriggerRegistration = nullptr;
+                    }
+                }
+            ) };
 
-                        if (name == backgroundTaskFullName)
-                        {
-                            s_pushTriggerRegistration = task.Value();
-                            return true;
-                        }
+            /* Every packaged app is already getting the benefit of COM Registration to handle notifications.
+               However, we would benefit from power management improvements in Background Infrastructure.
+               So we are continuing to register a Push trigger post 19041 builds */
+               
+            if (PushNotificationHelpers::IsPackagedAppScenario())
+            {
+                {
+                    auto lock = m_lock.lock_shared();
+                    THROW_HR_IF(E_INVALIDARG, m_pushTriggerRegistration);
+                }
 
-                        throw winrt::hresult_invalid_argument(L"RegisterActivator has different clsid registered.");
-                    }) };
+                winrt::hstring registeredClsidStr{ winrt::to_hstring(registeredClsid) };
+                winrt::hstring backgroundTaskFullName{ backgroundTaskName + registeredClsidStr };
 
+                bool isTaskRegistered = IsBackgroundTaskRegistered(backgroundTaskFullName);
+               
                 if (!isTaskRegistered)
                 {
+                    BackgroundTaskBuilder builder{ nullptr };
+
                     builder = BackgroundTaskBuilder();
                     builder.Name(backgroundTaskFullName);
 
                     PushNotificationTrigger trigger{};
                     builder.SetTrigger(trigger);
 
-                    THROW_HR_IF(E_NOTIMPL, !AppModel::Identity::IsPackagedProcess());
-
                     // In case the interface is not supported, let it throw.
```

Comment:
This comment is not needed. This is a PackagedAppScenario which by default indicates that the Cast below should pass. If it doesnt, it is unexpected and fatal.

Nothing to do with it not being supported.